libbpf-cargo/src/test.rs

```diff
@@ -1096,6 +1096,98 @@ fn test_skeleton_builder_deterministic() {
     assert_eq!(skel1, skel2);
 }
 
+/// Check that we generate a valid skeleton in the presence of duplicate
+/// structs in the BTF.
+#[test]
+fn test_skeleton_duplicate_struct() {
+    let (_dir, proj_dir, cargo_toml) = setup_temp_project();
+
+    create_dir(proj_dir.join("src/bpf")).expect("failed to create prog dir");
+    let prog1 = proj_dir.join("src/bpf/prog1.bpf.c");
+    write(
+        &prog1,
+        r#"
+            #include "vmlinux.h"
+
+            struct foobar {
+                int bar;
+            } foo;
+        "#,
+    )
+    .expect("failed to write prog.bpf.c");
+    let prog2 = proj_dir.join("src/bpf/prog2.bpf.c");
+    write(
+        &prog2,
+        r#"
+            #include "vmlinux.h"
+
+            typedef int nonsensicaltypedeftoint;
+
+            struct foobar {
+                nonsensicaltypedeftoint bar;
+            } baz;
+        "#,
+    )
+    .expect("failed to write prog.bpf.c");
+
+    add_vmlinux_header(&proj_dir);
+
+    let bpf_o = proj_dir.join("prog.bpf.o");
+    let _output = BpfObjBuilder::default()
+        .build_many([&prog1, &prog2], &bpf_o)
+        .unwrap();
+
+    let skel = proj_dir.join("src/bpf/skel.rs");
+    let () = SkeletonBuilder::new().obj(&bpf_o).generate(&skel).unwrap();
+
+    let mut cargo = OpenOptions::new()
+        .append(true)
+        .open(&cargo_toml)
+        .expect("failed to open Cargo.toml");
+
+    // Make test project use our development libbpf-rs version
+    writeln!(
+        cargo,
+        r#"
+        libbpf-rs = {{ path = "{}" }}
+        "#,
+        get_libbpf_rs_path().as_path().display()
+    )
+    .expect("failed to write to Cargo.toml");
+
+    let mut source = OpenOptions::new()
+        .write(true)
+        .truncate(true)
+        .open(proj_dir.join("src/main.rs"))
+        .expect("failed to open main.rs");
+
+    write!(
+        source,
+        r#"
+        #[path = "{skel_path}"]
+        mod skel;
+        use skel::*;
+
+        fn main() {{
+            let _foobar0 = types::foobar::default();
+            let _foobar1 = types::foobar_1::default();
```

Comment:
actually, wouldn't `foobar_2` make more sense (it's a second instance of the same name, we just don't explicitly mark the first one with `_1`)?

Reply:
I am living happily in my counting-starts-at-zero bubble :-)

But happy to change it.

Btw., strictly speaking I think the renaming algorithm is insufficient: we can't just assume that `<type>_<cnt>` is an available name, but should actually check that there is no collision with an existing type of that name.

Reply:
heh, I see :) I'd switch to two, but ultimately it doesn't matter all that much. If user gets to that point, there is no deterministic order, so whether it's foobar_1 or foobar_2, I wouldn't use any of those in the code directly because their ordering can change on the next compilation

and yeah, you are right about possibility of causing another unintended collision (which, on libbpf side, is mostly mitigated by using triple underscore, which makes it highly unlikely)